File: src/main/java/me/arasple/mc/trmenu/TrMenu.java
package me.arasple.mc.trmenu;

import com.google.common.collect.Lists;
import io.izzel.taboolib.loader.Plugin;
import io.izzel.taboolib.module.config.TConfig;
import io.izzel.taboolib.module.inject.TInject;
import me.arasple.mc.trmenu.menu.Menu;

import java.io.File;
import java.util.List;

/**
 * @author Arasple
 * @document https://trmenu.trixey.cn
 * @since 2019.10.2
 */
@Plugin.Version(5.15)
public final class TrMenu extends Plugin {

    @TInject(value = "settings.yml", locale = "LOCALE-PRIORITY")
    public static final TConfig SETTINGS = null;
    @TInject
    private static final TrMenu INSTANCE = null;
    @TInject(state = TInject.State.LOADING, init = "init", active = "active", cancel = "cancel")
    private static final TrMenuLoader LOADER = null;

    private static List<Menu> menus = Lists.newArrayList();

    public static List<Menu> getMenus() {
        return menus;
    }

    public static double getTrVersion() {
        return 1.19;
    }

    public static File getPluginFile() {
        return INSTANCE.getFile();
    }

}

File: src/main/kotlin/me/arasple/mc/trmenu/TrMenu.kt
package me.arasple.mc.trmenu

import me.arasple.mc.trmenu.module.conf.Loader
import me.arasple.mc.trmenu.module.conf.prop.RunningPerformance
import me.arasple.mc.trmenu.module.display.MenuSession
import me.arasple.mc.trmenu.module.internal.data.Metadata
import me.arasple.mc.trmenu.module.internal.hook.HookPlugin
import me.arasple.mc.trmenu.module.internal.inputer.Inputer.Companion.cancelWords
import me.arasple.mc.trmenu.module.internal.listener.ListenerItemInteract.interactCooldown
import me.arasple.mc.trmenu.module.internal.service.RegisterCommands
import me.arasple.mc.trmenu.module.internal.service.Shortcuts
import org.bukkit.Bukkit
import taboolib.common.platform.*
import taboolib.common.platform.function.console
import taboolib.module.configuration.Config
import taboolib.module.configuration.SecuredFile
import taboolib.module.lang.Language
import taboolib.module.lang.sendLang
import taboolib.platform.BukkitPlugin

/**
 * @author Arasple
 * @date 2021/1/24 9:51
 */
object TrMenu : Plugin() {

    @Config("settings.yml",true, autoReload = true)
    lateinit var SETTINGS: SecuredFile
        private set

    val plugin by lazy { BukkitPlugin.getInstance() }

    var performance = RunningPerformance.NORMAL
        private set
    
    override fun onLoad() {
        Language.default = "en_US"
        console().sendLang("Plugin-Loading", Bukkit.getVersion())
    }

    override fun onEnable() {
        SETTINGS.onReload { onSettingsReload() }
        onSettingsReload()
        Loader.loadMenus()
        Metadata.database
        console().sendLang("Plugin-Enabled", plugin.description.version)
        HookPlugin.printInfo()
    }

    override fun onDisable() {
        MenuSession.SESSIONS.entries.removeIf {
            it.value.close(closePacket = true, updateInventory = true)
            true
        }
    }

    private fun onSettingsReload() {
        performance = kotlin.runCatching {
            RunningPerformance.valueOf(SETTINGS.getString("Options.Running-Performance")!!)
        }.getOrNull() ?: RunningPerformance.NORMAL

        cancelWords.reload()
        interactCooldown.reload()
        Shortcuts.Type.load()
        RegisterCommands.load()
    }

}